README
Thread::Suspend version 1.22
============================

This module adds suspend and resume operations for threads.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires Perl 5.8.0 or later.

This module requires these other modules:

    threads version 1.39 or later
    threads::shared 1.01 or later
    Test::More 0.50 or later (for installation)

COPYRIGHT AND LICENCE

Copyright 2006 - 2009 Jerry D. Hedden <jdhedden AT cpan DOT org>

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

# EOF


Changes
Revision history for Perl extension Thread::Suspend.

1.22 Wed Nov 21 19:34:31 2012
	- Fix for test hangs on Windows

1.21 Fri Dec 24 17:58:16 2010
	- POD update

1.19 Thu Jun 12 13:46:30 2008
	- Fix SKIPs in tests

1.18 Tue Jun 10 16:58:13 2008
	- End all tests with exit(0)

1.17 Thu Feb 28 16:05:15 2008
	- More fixes to tests

1.16 Wed Feb 27 17:43:15 2008
	- More fixes to tests

1.15 Wed Feb 27 13:43:14 2008
	- More fixes to tests

1.14 Wed Feb 27 01:14:00 2008
	- More fixes to tests

1.13 Mon Feb 25 22:17:19 2008
	- Fixes to tests

1.12 Fri Feb 22 22:19:37 2008
	- Allow installation on non-threaded Perls

1.11 Wed Feb 20 17:26:30 2008
	- Build/test updates

1.09 Sat Oct 14 09:08:47 EST 2006
	- Check for $Config{useithreads}

1.08 Thu Oct 12 13:28:25 EDT 2006
	- Added example and POD tests

1.07 Wed Aug 30 14:25:35 EDT 2006
	- Support 5.8.0

1.06 Thu Aug 10 14:51:07 EDT 2006
	- More fixes to tests (sigh)

1.05 Wed Aug  2 13:54:17 EDT 2006
	- More fixes to tests

1.04 Tue Aug  1 13:56:38 EDT 2006
	- Fixes to tests

1.03 Thu Jul 13 08:53:36 EDT 2006
	- Added caveat to POD concerning locks

1.02 Wed Jul 12 16:00:03 EDT 2006
	- Fix lock issues in tests

1.01 Tue Jul 11 09:18:45 EDT 2006
	- Initial release



t/test.pl
use strict;
use warnings;


our @DONE :shared;

$SIG{'KILL'} = sub {
    $DONE[threads->tid()] = 1;
    threads->exit();
};


our %CHECKER :shared;

sub checker
{
    my $tid = threads->tid();
    while (1) {
        delete($CHECKER{$tid});
        threads->yield();
    }
}


sub pause
{
    threads->yield() for (0..$::nthreads);
    select(undef, undef, undef, shift);
    threads->yield() for (0..$::nthreads);
}

sub check {
    my ($thr, $state, $line) = @_;
    my $tid = $thr->tid();

    pause(0.1);
    delete($CHECKER{$tid});
    if (exists($CHECKER{$tid})) {
        ok(0, "BUG: \$CHECKER{$tid} not deleted");
    }
    $CHECKER{$tid} = $tid;

    if ($state eq 'running') {
        for (1..100) {
            pause(0.1);
            last if (! exists($CHECKER{$tid}));
        }
        ok(! exists($CHECKER{$tid}), "Thread $tid $state (line $line)");
    } else {
        for (1..3) {
            pause(0.1);
            last if (! exists($CHECKER{$tid}));
        }
        ok(exists($CHECKER{$tid}), "Thread $tid $state (line $line)");
    }
}


sub make_threads
{
    my $nthreads = shift;
    my @threads;
    push(@threads, threads->create('checker')) for (1..$nthreads);
    is(scalar(threads->list()), $nthreads, 'Threads created');
    return @threads;
}

1;


examples/suspend.pl
#!/usr/bin/perl

use strict;
use warnings;

$| = 1;

use threads;
use threads::shared;
use Thread::Suspend;

### Global Variables ###

# Flag to inform all threads that application is terminating
my $TERM :shared = 0;


### Signal Handling ###

# Gracefully terminate application on ^C
# or command line 'kill'
$SIG{'INT'} = $SIG{'TERM'} = sub { $TERM = 1; };


### Main Processing Section ###
MAIN:
{
    print("To terminate demo, hit ctrl-C\n\n");

    # Start the worker thread
    my $worker = threads->new('worker');
    $worker->detach();
    threads->yield();

    # Start the controller thread
    threads->new('controller', $worker)->detach();
    threads->yield();

    # Hang around until told to terminate
    sleep(1) until ($TERM);

    print("\e[?25h\n\e[2K\n\e[2K\n");   # Restore cursor
}

print("Done\n");
exit(0);


### Thread Entry Point Subroutines ###

# A worker thread
sub worker
{
    ### INITIALIZE ###

    printf("Working thread started and waiting.\n\n");
    threads->self()->suspend();

    ### WORK ###

    my $cnt = 0;
    while (! $TERM) {
        print("\rWorking: $cnt");
        $cnt++;
    }
}


sub controller
{
    my $worker = shift;    # The worker thread

    print("\n\e[ATo start worker thread, hit return: ");
    my $user = <STDIN>;

    while (! $TERM) {
        print("\e[A\e[KTo suspend worker thread, hit return...\n");
        print("\e[3A\e[K\e[?25l");   # Hide cursor
        $worker->resume();
        $user = <STDIN>;
        $worker->suspend();
        threads->yield();
        print("\e[2K\e[?25h\n");     # Restore cursor
        print("\e[KTo resume worker thread, hit return: ");
        $user = <STDIN>;
    }
}

__END__

=head1 NAME

suspend.pl - Simple example illustrating threads suspend and resume operations

=head1 DESCRIPTION

A simplistic example with one thread controlling the execution of another
using suspend and resume operations.

=head1 SEE ALSO

L<threads> and L<threads::shared>

=head1 AUTHOR

Jerry D. Hedden, S<E<lt>jdhedden AT cpan DOT orgE<gt>>

=head1 COPYRIGHT AND LICENSE

Copyright 2006 - 2007 Jerry D. Hedden. All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


lib/Thread/Suspend.pm
package Thread::Suspend; {

use strict;
use warnings;

our $VERSION = '1.22';

use threads 1.39;
use threads::shared 1.01;

my %SUSPEND :shared;    # Thread suspension counts by TID

my $SIGNAL = 'STOP';    # Default suspension signal


sub import
{
    my $class = shift;   # Not used

    # Set the signal for suspend operations
    while (my $sig = shift) {
        $SIGNAL = $sig;
    }
    $SIGNAL =~ s/^SIG//;

    # Set up the suspend signal handler
    $SIG{$SIGNAL} = sub {
        my $tid = threads->tid();
        lock(%SUSPEND);
        while ($SUSPEND{$tid}) {
            cond_wait(%SUSPEND);
        }
    };
}


sub threads::suspend
{
    my ($thing, @threads) = @_;

    if ($thing eq 'threads') {
        if (@threads) {
            # Suspend specified list of threads
            @threads = grep { $_ }
                       map  { (ref($_) eq 'threads')
                                    ? $_
                                    : threads->object($_) }
                            @threads;
        } else {
            # Suspend all non-detached threads
            push(@threads, threads->list(threads::running));
        }
    } else {
        # Suspend a single thread
        push(@threads, $thing);
    }

    # Suspend threads
    lock(%SUSPEND);
    foreach my $thr (@threads) {
        my $tid = $thr->tid();
        # Increment suspension count
        if (! $SUSPEND{$tid}++) {
            # Send suspend signal if not currently suspended
            $thr->kill($SIGNAL);
            if (! $thr->is_running()) {
                # Thread terminated before it could be suspended
                delete($SUSPEND{$tid});
            }
        }
    }

    # Return list of affected threads
    return ($thing eq 'threads')
                    ? grep { $_->is_running() } @threads
                    : $thing;
}


sub threads::resume
{
    my ($thing, @threads) = @_;

    lock(%SUSPEND);
    if ($thing eq 'threads') {
        if (@threads) {
            # Resume specified threads
            @threads = grep { $_ }
                       map  { (ref($_) eq 'threads')
                                    ? $_
                                    : threads->object($_) }
                            @threads;
        } else {
            # Resume all threads
            @threads = grep { $_ }
                       map  { threads->object($_) }
                            keys(%SUSPEND);
        }
    } else {
        # Resume a single thread
        push(@threads, $thing);
    }

    # Resume threads
    my $resume = 0;
    foreach my $thr (@threads) {
        my $tid = $thr->tid();
        if ($SUSPEND{$tid}) {
            # Decrement suspension count
            if (! --$SUSPEND{$tid}) {
                # Suspension count reached zero
                $resume = 1;
                delete($SUSPEND{$tid});
            }
        }
    }
    # Broadcast any resumptions
    if ($resume) {
        cond_broadcast(%SUSPEND);
    }

    # Return list of affected threads
    return ($thing eq 'threads') ? @threads : $thing;
}


sub threads::is_suspended
{
    my $item = shift;

    lock(%SUSPEND);
    if ($item eq 'threads') {
        # Return list of all non-detached suspended threads
        return (grep { $_ }
                map  { threads->object($_) }
                    keys(%SUSPEND));

    } else {
        # Return suspension count for a single thread
        my $tid = $item->tid();
        return ($SUSPEND{$tid}) ? $SUSPEND{$tid} : 0;
    }
}

}

1;

__END__

=head1 NAME

Thread::Suspend - Suspend and resume operations for threads

=head1 VERSION

This document describes Thread::Suspend version 1.22

=head1 SYNOPSIS

    use Thread::Suspend 'SIGUSR1';      # Set the suspension signal
    use Thread::Suspend;                #  Defaults to 'STOP'

    $thr->suspend();                    # Suspend a thread
    threads->suspend();                 # Suspend all non-detached threads
    threads->suspend($thr, $tid, ...);  # Suspend multiple threads using
                                        #   objects or TIDs

    $thr->is_suspended();               # Returns suspension count
    threads->is_suspended();            # Returns list of all suspended threads

    $thr->resume();                     # Resume a thread
    threads->resume();                  # Resume all threads
    threads->resume($thr, $tid, ...);   # Resume multiple threads

=head1 DESCRIPTION

This module adds suspend and resume operations for threads.

Suspensions are cumulative, and need to be matched by an equal number of
resume calls.

=head2 Declaration

This module must be imported prior to any threads being created.

Suspension is accomplished via a signal handler which is used by all threads
on which suspend operations are performed.  The signal for this operation can
be specified when this module is declared, and defaults to C<SIGSTOP>.
Consequently, the application and its threads must not specify some other
handler for use with the suspend signal.

=over

=item use Thread::Suspend;

Declares this module, and defaults to using C<SIGSTOP> for suspend operations.

=item use Thread::Suspend 'SIGUSR1';

=item use Thread::Suspend 'Signal' => 11;

Declares this module, and uses the specified signal for suspend operations.
Signals may be specified by the same names or (positive) numbers as supported
by L<kill()|perlfunc/"kill SIGNAL, LIST">.

=back

=head2 Methods

=over

=item $thr->suspend()

Adds 1 to the suspension count of the thread, and suspends its execution if
running.  Returns the I<threads> object.

It is possible for a thread to suspend itself.  This is useful for starting a
thread early in an application, and having it C<wait> until needed:

    sub thr_func
    {
        # Suspend until needed
        threads->self()->suspend();
        ...
    }

=item threads->suspend()

Adds 1 to the suspension count of all non-detached threads, and
suspends their execution if running.  Returns a list of those threads.

=item threads->suspend($thr, $tid, ...)

Adds 1 to the suspension count of the threads specified by their objects or
TIDs (for non-detached threads), and suspends their execution if running.
Returns a list of the corresponding I<threads> objects affected by the call.

=item $thr->is_suspended()

Returns the suspension count for the thread.

=item threads->is_suspended()

Returns a list of currently suspended, non-detached threads.

=item $thr->resume()

Decrements the suspension count for a thread.  The thread will resume
execution if the count reaches zero.  Returns the I<threads> object.

=item threads->resume()

Decrements the suspension count for all currently suspended, non-detached
threads.  Those threads that reach a count of zero will resume execution.
Returns a list of the threads operated on.

Given possible multiple levels of suspension, you can ensure that all
(non-detached) threads are running using:

    while (threads->resume()) { }

=item threads->resume($thr, $tid, ...)

Decrements the suspension count of the threads specified by their objects or
TIDs (for non-detached threads).  Those threads that reach a count of zero
will resume execution.  Returns a list of the threads operated on.

=back

=head1 CAVEATS

Subject to the limitations of L<threads/"THREAD SIGNALLING">.

A thread that has been suspended will not respond to any other signals or
commands until its suspension count is brought back to zero via resume calls.

Any locks held by a thread when it is suspended will remain in effect.  To
alleviate this potential problem, lock any such variables as part of a limited
scope that also contains the suspension call:

    {
        lock($var);
        $thr->suspend();
    }

Calling C<-E<gt>resume()> on an non-suspended thread is ignored.

Detached threads can only be operated upon if their I<threads> object is used.
For example, the following works:

    my $thr = threads->create(...);
    $thr->detach();
    ...
    $thr->suspend();  # or threads->suspend($thr);
    ...
    $thr->resume();   # or threads->resume($thr);

Threads that have finished execution are, for the most part, ignored by this
module.

=head1 REQUIREMENTS

Perl 5.8.0 or later

L<threads> 1.39 or later

L<threads::shared> 1.01 or later

L<Test::More> 0.50 or later (for installation)

=head1 SEE ALSO

Thread::Suspend Discussion Forum on CPAN:
L<http://www.cpanforum.com/dist/Thread-Suspend>

L<threads>, L<threads::shared>

=head1 AUTHOR

Jerry D. Hedden, S<E<lt>jdhedden AT cpan DOT orgE<gt>>

=head1 COPYRIGHT AND LICENSE

Copyright 2006 - 2009 Jerry D. Hedden. All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


Makefile.PL
# Module makefile for Thread::Suspend (using ExtUtils::MakeMaker)

use strict;
use warnings;

use ExtUtils::MakeMaker;

# Construct make file
WriteMakefile(
    'NAME'          => 'Thread::Suspend',
    'AUTHOR'        => 'Jerry D. Hedden <jdhedden AT cpan DOT org>',
    'VERSION_FROM'  => 'lib/Thread/Suspend.pm',
    'ABSTRACT_FROM' => 'lib/Thread/Suspend.pm',
    'PREREQ_PM'     => {
        'threads'         => 1.39,
        'threads::shared' => 1.01,
        'Test::More'      => 0.50,
    },

    ((ExtUtils::MakeMaker->VERSION() lt '6.25') ?
        ('PL_FILES' => { })            : ()),
    ((ExtUtils::MakeMaker->VERSION() gt '6.30') ?
        ('LICENSE'  => 'perl')         : ()),
);

# Additional 'make' targets
sub MY::postamble
{
    return <<'_EXTRAS_';
fixfiles:
	@dos2unix `cat MANIFEST`
	@$(CHMOD) 644 `cat MANIFEST`
	@$(CHMOD) 755 examples/*.pl
_EXTRAS_
}

# EOF
